pkg/bench/workload.go

```diff
@@ -194,6 +209,91 @@ func (w *writeWorkload) generateTimeSeries(id string, t time.Time) []prompb.Time
 	return timeseries
 }
 
+type batchReq struct {
+	batch   []prompb.TimeSeries
+	wg      *sync.WaitGroup
+	putBack chan []prompb.TimeSeries
+}
+
+func (w *writeWorkload) getSeriesBuffer(ctx context.Context) []prompb.TimeSeries {
+	select {
+	case <-ctx.Done():
+		return nil
+	case seriesBuffer := <-w.seriesBufferChan:
+		return seriesBuffer[:0]
+	}
+}
+
+func (w *writeWorkload) generateWriteBatch(ctx context.Context, id string, numBuffers int, seriesChan chan batchReq) error {
+	w.seriesBufferChan = make(chan []prompb.TimeSeries, numBuffers)
+	for i := 0; i < numBuffers; i++ {
+		w.seriesBufferChan <- make([]prompb.TimeSeries, 0, w.options.BatchSize)
+	}
+
+	seriesBuffer := w.getSeriesBuffer(ctx)
+	ticker := time.NewTicker(w.options.Interval)
+
+	defer close(seriesChan)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case timeNow := <-ticker.C:
+			now := timeNow.UnixNano() / int64(time.Millisecond)
+			wg := &sync.WaitGroup{}
+			for replicaNum := 0; replicaNum < w.replicas; replicaNum++ {
+				replicaLabel := prompb.Label{Name: "bench_replica", Value: fmt.Sprintf("replica-%05d", replicaNum)}
+				idLabel := prompb.Label{Name: "bench_id", Value: id}
+				for _, series := range w.series {
+					var value float64
+					switch series.seriesType {
+					case GaugeZero:
+						value = 0
+					case GaugeRandom:
+						value = rand.Float64()
+					case CounterOne:
+						value = series.lastValue + 1
+					case CounterRandom:
+						value = series.lastValue + float64(rand.Int())
+					default:
+						return fmt.Errorf("unknown series type %v", series.seriesType)
+					}
+					series.lastValue = value
+					for _, labelSet := range series.labelSets {
+						if len(seriesBuffer) == w.options.BatchSize {
+							wg.Add(1)
+							seriesChan <- batchReq{seriesBuffer, wg, w.seriesBufferChan}
+							seriesBuffer = w.getSeriesBuffer(ctx)
+						}
+						newLabelSet := make([]prompb.Label, len(labelSet)+2)
+						copy(newLabelSet, labelSet)
+
+						newLabelSet[len(newLabelSet)-2] = replicaLabel
+						newLabelSet[len(newLabelSet)-1] = idLabel
+						seriesBuffer = append(seriesBuffer, prompb.TimeSeries{
+							Labels: newLabelSet,
+							Samples: []prompb.Sample{{
+								Timestamp: now,
+								Value:     value,
+							}},
+						})
+					}
+				}
+			}
+			if len(seriesBuffer) > 0 {
+				wg.Add(1)
+				seriesChan <- batchReq{seriesBuffer, wg, w.seriesBufferChan}
+				seriesBuffer = w.getSeriesBuffer(ctx)
+			}
+			wg.Wait()
+			if time.Since(timeNow) > w.options.Interval {
+				w.missedIterations.Inc()
```

Comment:
do I understand this correctly that you're incrementing this whenever you finish late, to indicate that the _next_ iteration is starting late?

Reply:
Yea, it's a decent indicator as to whether the write load is actually being delivered or if it's falling behind.